libfreerdp/core/nla.c

```diff
@@ -1936,110 +1899,66 @@ SECURITY_STATUS nla_decrypt_public_key_hash(rdpNla* nla)
 	return status;
 }
 
-BOOL nla_read_ts_password_creds(rdpNla* nla, wStream* s)
+static BOOL nla_read_ts_credentials(rdpNla* nla, SecBuffer* data, size_t offset)
 {
-	size_t length;
-	size_t userLen = 0;
-	size_t domainLen = 0;
-	size_t passwordLen = 0;
-	const WCHAR* user = NULL;
-	const WCHAR* domain = NULL;
-	const WCHAR* password = NULL;
+	WinPrAsn1Decoder dec, dec2;
+	WinPrAsn1_OctetString credentials;
+	BOOL error;
+	WinPrAsn1_INTEGER credType;
 
 	WINPR_ASSERT(nla);
-	WINPR_ASSERT(s);
-
-	if (!nla->identity)
-	{
-		WLog_ERR(TAG, "nla->identity is NULL!");
-		return FALSE;
-	}
-
-	/* TSPasswordCreds (SEQUENCE)
-	 * Initialise to default values. */
-
-	sspi_FreeAuthIdentity(nla->identity);
-
-	if (!ber_read_sequence_tag(s, &length))
-		return FALSE;
-
-	/* The sequence is empty, return early,
-	 * TSPasswordCreds (SEQUENCE) is optional. */
-	if (length == 0)
-		return TRUE;
-
-	/* [0] domainName (OCTET STRING) */
-	if (!ber_read_contextual_tag(s, 0, &length, TRUE) || !ber_read_octet_string_tag(s, &length))
-	{
+	if (!data || data->cbBuffer < offset)
 		return FALSE;
-	}
 
-	domainLen = length / sizeof(WCHAR);
-	if (length > 0)
-		domain = Stream_PointerAs(s, const WCHAR);
+	WinPrAsn1Decoder_InitMem(&dec, WINPR_ASN1_DER, (BYTE*)data->pvBuffer + offset,
+	                         data->cbBuffer - offset);
 
-	if (!Stream_SafeSeek(s, length))
+	/* TSCredentials */
+	if (!WinPrAsn1DecReadSequence(&dec, &dec2))
 		return FALSE;
+	dec = dec2;
 
-	/* [1] userName (OCTET STRING) */
-	if (!ber_read_contextual_tag(s, 1, &length, TRUE) || !ber_read_octet_string_tag(s, &length))
+	/* credType [0] INTEGER */
+	if (!WinPrAsn1DecReadContextualInteger(&dec, 0, &error, &credType))
 		return FALSE;
 
-	userLen = length / sizeof(WCHAR);
-	if (length > 0)
-		user = Stream_PointerAs(s, const WCHAR);
-
-	if (!Stream_SafeSeek(s, length))
+	/* credentials [1] OCTET STRING */
+	if (!WinPrAsn1DecReadContextualOctetString(&dec, 1, &error, &credentials, FALSE))
 		return FALSE;
 
-	/* [2] password (OCTET STRING) */
-	if (!ber_read_contextual_tag(s, 2, &length, TRUE) || !ber_read_octet_string_tag(s, &length))
-		return FALSE;
+	WinPrAsn1Decoder_InitMem(&dec, WINPR_ASN1_DER, credentials.data, credentials.len);
 
-	passwordLen = length / sizeof(WCHAR);
-	if (length > 0)
-		password = Stream_PointerAs(s, const WCHAR);
-
-	if (!Stream_SafeSeek(s, length))
-		return FALSE;
+	if (credType == 1)
```

Comment:
If we receive some `TSSmartCardCreds` or some `TSRemoteGuardCreds` we return TRUE even if we've read nothing, perhaps you can change the `if (credType==1) {` by a `switch(credType) { case 1: <code>... }` and have a default statement that returns FALSE until we have implemented the other kind of credentials

Reply:
nla has no "auto-login", either the transport is established (and authenticated), or it fails (no login greeter in NLA).

Reply:
Hum you're perfectly correct.